src/datatypes/src/vectors/decimal.rs

```diff
@@ -0,0 +1,465 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+use std::fmt::Debug;
+use std::sync::Arc;
+
+use arrow::array::ArrayData;
+use arrow_array::builder::{ArrayBuilder, Decimal128Builder};
+use arrow_array::iterator::ArrayIter;
+use arrow_array::{Array, ArrayRef, Decimal128Array};
+use common_decimal::Decimal128;
+use snafu::{OptionExt, ResultExt};
+
+use super::{MutableVector, Validity, Vector, VectorRef};
+use crate::arrow::datatypes::DataType as ArrowDataType;
+use crate::data_type::ConcreteDataType;
+use crate::error::{self, CastTypeSnafu, InvalidArgumentsSnafu, Result};
+use crate::prelude::{ScalarVector, ScalarVectorBuilder};
+use crate::serialize::Serializable;
+use crate::value::{Value, ValueRef};
+use crate::vectors;
+
+/// Decimal128Vector is a vector keep i128 values with precision and scale.
+#[derive(Debug, PartialEq)]
+pub struct Decimal128Vector {
+    array: Decimal128Array,
+}
+
+impl Decimal128Vector {
+    pub fn new(array: Decimal128Array) -> Self {
+        Self { array }
+    }
+
+    pub fn from_array_data(array: ArrayData) -> Self {
+        Self {
+            array: Decimal128Array::from(array),
+        }
+    }
+
+    /// Construct Vector from i128 values
+    pub fn from_values<I: IntoIterator<Item = i128>>(iter: I) -> Self {
+        Self {
+            array: Decimal128Array::from_iter_values(iter),
+        }
+    }
+
+    pub fn from_slice<P: AsRef<[i128]>>(slice: P) -> Self {
+        let iter = slice.as_ref().iter().copied();
+        Self {
+            array: Decimal128Array::from_iter_values(iter),
+        }
+    }
+
+    pub fn from_wrapper_slice<P: AsRef<[Decimal128]>>(slice: P) -> Self {
+        let iter = slice.as_ref().iter().copied().map(|v| v.val());
+        Self {
+            array: Decimal128Array::from_iter_values(iter),
+        }
+    }
+
+    pub fn to_array_data(&self) -> ArrayData {
+        self.array.to_data()
+    }
```

Comment:
We don't need to use ArrayData now. Arrow's array already supports clone so we can clone the inner array. We can remove `to_array_data()/from_array_data()`